### DISCUSSION TITLE
After upgrading from version 7.7.5 to 7.8.5, the wave graph is not rendered

### DISCUSSION THREAD

**Discussion by TD-jizhijun** [katspaugh/wavesurfer.js#3865] [Q&A]
Hi team, after upgrading from version 7.7.5 to 7.8.5, the wave graph is not rendered. I don't know where the problem is, could you give me some guidance?
Code details:

import React, { CSSProperties, useEffect, useRef, useState } from 'react';
import Box from '@cobalt/react-box';
import { useTheme } from '@cobalt/react-theme-provider';
import AudioPart from 'src/MediaPlayer/audio/AudioPart';
import styles from 'src/UnifiedPlayer/styles/helpers.module.scss';
import { waveReady } from 'src/UnifiedPlayer/utils/wave-utils';
import WaveSurfer from 'wavesurfer.js';

const WaveElementId = 'waveElementId';

const WAVE_HEIGHT = 50;
const BAR_WIDTH = 4;
const BAR_HEIGHT = 0.7;
const BAR_GAP = 2;

type Props = {
  onSeek: (absoluteSeekPosition: number) => void;
  data: {
    audioParts: AudioPart[];
  };
  hasCookieAuthentication?: boolean;
};

const Waves: React.FC<Props> = ({
  onSeek,
  data,
  hasCookieAuthentication = false,
}) => {
  const audiosLoaded = useRef(false);
  const [audioParts, setAudioParts] = useState<AudioPart[]>([]);
  const [waves, setWaves] = useState<WaveSurfer[]>([]);
  const [failedIndexes, setFailedIndexes] = useState<number[]>([]);
  const theme = useTheme();

  const MIDNIGHT_400 = theme.midnight400;
  const MIDNIGHT_700 = theme.midnight700;

  if (data.audioParts !== audioParts) {
    setAudioParts(data.audioParts);
  }

  const onDestroy = () => {
    if (waves?.length) {
      // destroy WaveSurfer instances
      waves.forEach((wave) => {
        wave.unAll();
        wave.destroy();
      });

      // reset waves references and time grid divisions
      setWaves([]);
      setFailedIndexes([]);
      audiosLoaded.current = false;
    }
  };

  useEffect(() => {
    onDestroy();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [audioParts]);

  useEffect(() => {
    // wait for wave widths
    if (!audioParts.length) {
      return;
    }

    // eslint-disable-next-line no-console
    console.log(111, audioParts);

    const waveReadyPromises: Promise<void>[] = [];
    const failed: number[] = [];

    const channelOptions = {
      waveColor: MIDNIGHT_400,
      progressColor: MIDNIGHT_700,
      cursorColor: MIDNIGHT_700,
      barWidth: BAR_WIDTH,
      barGap: BAR_GAP,
      barAlign: 'bottom' as const,
    };

    const wavesurferOptions = {
      height: WAVE_HEIGHT,
      barHeight: BAR_HEIGHT,
      splitChannels: [channelOptions, channelOptions],
      normalize: false,
      cursorWidth: 3,
      fillParent: true,
      hideScrollbar: true,
    };

    const newWaves = audioParts.map((part, index) => {
      const wave = WaveSurfer.create({
        ...wavesurferOptions,
        container: `#${WaveElementId}_${index}`,
        ...(hasCookieAuthentication && {
          fetchParams: {
            cache: 'default',
            mode: 'cors',
            method: 'GET',
            credentials: 'include',
          },
        }),
      });

      // when wave resolved, add to ready Promises so we will have a global ready perception
      // eslint-disable-next-line fp/no-mutating-methods
      waveReadyPromises.push(
        waveReady(wave).catch(() => {
          if (!failed.includes(index)) {
            // eslint-disable-next-line fp/no-mutating-methods
            failed.push(index);
          }
        })
      );

      wave.setMediaElement(part.audio);

      return wave;
    });

    // after all loads, update loading flag
    Promise.all(waveReadyPromises).then(() => {
      setWaves(newWaves);
      setFailedIndexes(failed);
      audiosLoaded.current = true;
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [audioParts]);

  // Watch for waves array change so we must subscribe to seek events
  useEffect(() => {
    if (!waves.length || !audiosLoaded.current) {
      return;
    }

    const failedParts = audioParts.filter((_, index) =>
      failedIndexes.includes(index)
    );

    waves.forEach((wave) => {
      wave.unAll();
      wave.on('click', (value: number) => {
        // get the current audio part and corresponding index containing current audio
        const currentPartIndex = audioParts.findIndex(
          (part) => part.audio === wave.getMediaElement()
        );

        const currentPart = audioParts[currentPartIndex];

        // for failed audio (not in the filtered array) there is no seek
        if (failedParts.includes(currentPart)) {
          return;
        }

        onSeek(currentPart.relativeTime + currentPart.duration * value);
      });
    });
  }, [waves, onSeek, audioParts, failedIndexes]);

  const wavesContainerStyle = {
    backgroundColor: theme.midnight200,
    height: WAVE_HEIGHT * 2,
  } as CSSProperties;

  const audioTotaltime = audioParts.reduce(
    (prev, curr) => (curr.audio?.duration || curr.duration) + prev,
    0
  );

  return (
    <Box className={styles.wavesContainer} style={wavesContainerStyle}>
      {audioParts.map((audioPart, index) => {
        const { audio }: { audio: HTMLAudioElement } = audioPart;

        const divWidth = audioPart.duration / audioTotaltime;

        return (
          <Box
            key={`gridWave-${audio?.src}`}
            className={`${WaveElementId} ${
              failedIndexes.includes(index) ? styles.failedWave : ''
            } ${styles.waveElement}`}
            data-duration={audio.duration > 0 ? audio.duration : 0}
            data-qa={`${WaveElementId}_${index}`}
            id={`${WaveElementId}_${index}`}
            width={`${divWidth * 100}%`}
          />
        );
      })}
    </Box>
  );
};

export default Waves;

Before:
![image](https://github.com/user-attachments/assets/33223ece-59c8-45c5-a533-c567588d5212)

After:
![image](https://github.com/user-attachments/assets/3598cc3b-040f-4bfe-a986-ef777f092ca1)

**Comment by katspaugh** [comment]
7.8.0 introduced lazy rendering. Only the currently visible portion of the waveform is rendered. This change didn’t introduce any public API changes. Make sure your code doesn’t rely on any internal elements.

**Comment by TD-jizhijun** [comment]
```
// A basic example

import WaveSurfer from 'wavesurfer.js'

const wavesurfer = WaveSurfer.create({
  container: document.body,
  waveColor: 'rgb(200, 0, 200)',
  progressColor: 'rgb(100, 0, 100)',
  barHeight: 0.7,
  splitChannels: [
    {
      barWidth: 2,
      barGap: 2,
      barAlign: 'bottom',
    },
    {
      barWidth: 2,
      barGap: 2,
      barAlign: 'bottom',
    },
  ]
})
wavesurfer.once('ready', () => {
  wavesurfer.setMediaElement('<audio preload="auto" crossorigin="" src="blob:http://localhost:3000/9dfbddcc-c654-49b1-8926-d913f2570b0d"></audio>')
})
wavesurfer.on('click', () => {
  wavesurfer.play()
})
```
I made some changes in your project case and found that the waveform was not rendered. This is similar to what I implemented above. Do you know where the problem is?

**Reply by katspaugh** [reply]
Why do you expect [setMediaElement](https://wavesurfer.xyz/docs/classes/wavesurfer.default#setMediaElement) to accept a string?

**Comment by TD-jizhijun** [comment]
I checked and printed the corresponding data, and they all have values, but the page is not rendered. I don't know why? The lazy render you mentioned above, for my code logic here, does it mean that I must set the URL first when creating it before loading the waveform?
![image](https://github.com/user-attachments/assets/7cb357c2-55f9-4f2e-8fed-dd01cdf5a887)

**Reply by katspaugh** [reply]
I told you above, you're loading audio wrong. You're not loading any audio in fact.

**Comment by TD-jizhijun** [comment]
> Hi team, after upgrading from version 7.7.5 to 7.8.5, the wave graph is not rendered. I don't know where the problem is, could you give me some guidance? Code details:
> 
> import React, { CSSProperties, useEffect, useRef, useState } from 'react'; import Box from '@cobalt/react-box'; import { useTheme } from '@cobalt/react-theme-provider'; import AudioPart from 'src/MediaPlayer/audio/AudioPart'; import styles from 'src/UnifiedPlayer/styles/helpers.module.scss'; import { waveReady } from 'src/UnifiedPlayer/utils/wave-utils'; import WaveSurfer from 'wavesurfer.js';
> 
> const WaveElementId = 'waveElementId';
> 
> const WAVE_HEIGHT = 50; const BAR_WIDTH = 4; const BAR_HEIGHT = 0.7; const BAR_GAP = 2;
> 
> type Props = { onSeek: (absoluteSeekPosition: number) => void; data: { audioParts: AudioPart[]; }; hasCookieAuthentication?: boolean; };
> 
> const Waves: React.FC = ({ onSeek, data, hasCookieAuthentication = false, }) => { const audiosLoaded = useRef(false); const [audioParts, setAudioParts] = useState<AudioPart[]>([]); const [waves, setWaves] = useState<WaveSurfer[]>([]); const [failedIndexes, setFailedIndexes] = useState<number[]>([]); const theme = useTheme();
> 
> const MIDNIGHT_400 = theme.midnight400; const MIDNIGHT_700 = theme.midnight700;
> 
> if (data.audioParts !== audioParts) { setAudioParts(data.audioParts); }
> 
> const onDestroy = () => { if (waves?.length) { // destroy WaveSurfer instances waves.forEach((wave) => { wave.unAll(); wave.destroy(); });
> 
> ```
>   // reset waves references and time grid divisions
>   setWaves([]);
>   setFailedIndexes([]);
>   audiosLoaded.current = false;
> }
> ```
> 
> };
> 
> useEffect(() => { onDestroy(); // eslint-disable-next-line react-hooks/exhaustive-deps }, [audioParts]);
> 
> useEffect(() => { // wait for wave widths if (!audioParts.length) { return; }
> 
> ```
> // eslint-disable-next-line no-console
> console.log(111, audioParts);
> 
> const waveReadyPromises: Promise<void>[] = [];
> const failed: number[] = [];
> 
> const channelOptions = {
>   waveColor: MIDNIGHT_400,
>   progressColor: MIDNIGHT_700,
>   cursorColor: MIDNIGHT_700,
>   barWidth: BAR_WIDTH,
>   barGap: BAR_GAP,
>   barAlign: 'bottom' as const,
> };
> 
> const wavesurferOptions = {
>   height: WAVE_HEIGHT,
>   barHeight: BAR_HEIGHT,
>   splitChannels: [channelOptions, channelOptions],
>   normalize: false,
>   cursorWidth: 3,
>   fillParent: true,
>   hideScrollbar: true,
> };
> 
> const newWaves = audioParts.map((part, index) => {
>   const wave = WaveSurfer.create({
>     ...wavesurferOptions,
>     container: `#${WaveElementId}_${index}`,
>     ...(hasCookieAuthentication && {
>       fetchParams: {
>         cache: 'default',
>         mode: 'cors',
>         method: 'GET',
>         credentials: 'include',
>       },
>     }),
>   });
> 
>   // when wave resolved, add to ready Promises so we will have a global ready perception
>   // eslint-disable-next-line fp/no-mutating-methods
>   waveReadyPromises.push(
>     waveReady(wave).catch(() => {
>       if (!failed.includes(index)) {
>         // eslint-disable-next-line fp/no-mutating-methods
>         failed.push(index);
>       }
>     })
>   );
> 
>   wave.setMediaElement(part.audio);
> 
>   return wave;
> });
> 
> // after all loads, update loading flag
> Promise.all(waveReadyPromises).then(() => {
>   setWaves(newWaves);
>   setFailedIndexes(failed);
>   audiosLoaded.current = true;
> });
> // eslint-disable-next-line react-hooks/exhaustive-deps
> ```
> 
> }, [audioParts]);
> 
> // Watch for waves array change so we must subscribe to seek events useEffect(() => { if (!waves.length || !audiosLoaded.current) { return; }
> 
> ```
> const failedParts = audioParts.filter((_, index) =>
>   failedIndexes.includes(index)
> );
> 
> waves.forEach((wave) => {
>   wave.unAll();
>   wave.on('click', (value: number) => {
>     // get the current audio part and corresponding index containing current audio
>     const currentPartIndex = audioParts.findIndex(
>       (part) => part.audio === wave.getMediaElement()
>     );
> 
>     const currentPart = audioParts[currentPartIndex];
> 
>     // for failed audio (not in the filtered array) there is no seek
>     if (failedParts.includes(currentPart)) {
>       return;
>     }
> 
>     onSeek(currentPart.relativeTime + currentPart.duration * value);
>   });
> });
> ```
> 
> }, [waves, onSeek, audioParts, failedIndexes]);
> 
> const wavesContainerStyle = { backgroundColor: theme.midnight200, height: WAVE_HEIGHT * 2, } as CSSProperties;
> 
> const audioTotaltime = audioParts.reduce( (prev, curr) => (curr.audio?.duration || curr.duration) + prev, 0 );
> 
> return ( {audioParts.map((audioPart, index) => { const { audio }: { audio: HTMLAudioElement } = audioPart;
> 
> ```
>     const divWidth = audioPart.duration / audioTotaltime;
> 
>     return (
>       <Box
>         key={`gridWave-${audio?.src}`}
>         className={`${WaveElementId} ${
>           failedIndexes.includes(index) ? styles.failedWave : ''
>         } ${styles.waveElement}`}
>         data-duration={audio.duration > 0 ? audio.duration : 0}
>         data-qa={`${WaveElementId}_${index}`}
>         id={`${WaveElementId}_${index}`}
>         width={`${divWidth * 100}%`}
>       />
>     );
>   })}
> </Box>
> ```
> 
> ); };
> 
> export default Waves;
> 
> Before: ![image](https://private-user-images.githubusercontent.com/91597937/366291116-33223ece-59c8-45c5-a533-c567588d5212.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNDQxNjYsIm5iZiI6MTcyNjA0Mzg2NiwicGF0aCI6Ii85MTU5NzkzNy8zNjYyOTExMTYtMzMyMjNlY2UtNTljOC00NWM1LWE1MzMtYzU2NzU4OGQ1MjEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTExVDA4Mzc0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiYzUzN2Q3ZjM2NTZlNGM2YWIzYTRjZTdkNDc0ZmRiZTY4OWRlODBkYzBjOWZiNTNiOTk4ZGJhOTJlMDM3MDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tTQeqsZUgB3QW_j3sfUThbpVuSLuob7U9FVJIux6LpQ)
> 
> After: ![image](https://private-user-images.githubusercontent.com/91597937/366291186-3598cc3b-040f-4bfe-a986-ef777f092ca1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNDQxNjYsIm5iZiI6MTcyNjA0Mzg2NiwicGF0aCI6Ii85MTU5NzkzNy8zNjYyOTExODYtMzU5OGNjM2ItMDQwZi00YmZlLWE5ODYtZWY3NzdmMDkyY2ExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTExVDA4Mzc0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlZTQ5NGM1Njk5OGU4MGZkYmUzZGI5YmQ1MzZjZjBkZWQxZDZlM2Y4MmYxNzY0N2FhMWUzM2Y4MjQ0M2RjN2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ckgjTUwHRprMR_xU5c3valxUe01vrAMO0ZiQbLnXB00)

In my own actual code, audio is the correct audio element, because the function was normal before the upgrade, but it was not rendered after the upgrade.
![image](https://github.com/user-attachments/assets/cdf16a58-4e16-433f-8f05-32f211d66e61)

**Reply by TD-jizhijun** [reply]
![image](https://github.com/user-attachments/assets/9dd728aa-1d4d-4d9c-9587-75e717be82cb)

**Comment by TD-jizhijun** [comment]
Hi @katspaugh, I have found the problem. The update of version 7.7.15 seems to affect the setMediaElement method. I don't know if it is a bug, but I have solved it by passing the value through the media field when creating. You can verify whether setMediaElement is available when you have time. At present, I found that waveform rendering cannot be performed using this method.

**Reply by katspaugh** [reply]
I’ll check, thanks!
